File: sorting/IntermediateSort/QuickSort.js
'use strict';
// Quick Sort

// Pivot Helper

/**
 * In order to 
 */

// const pivot = (arr, start=0, end = arr.length -1 )=>{


    
//     const fistElement = arr[start]
//     let positionIndex = 1;
//     for(let i = 1; i < arr.length; i++){

//         if(fistElement > arr[i]){
//             [arr[positionIndex], arr[i]] = [arr[i], arr[positionIndex]]
//             positionIndex++
//         }

//     }
// console.log(positionIndex);
//   [arr[start], arr[positionIndex -1]] = [arr[positionIndex -1], arr[start]]

//     return arr

    
// }


// const arr = [5, 2, 1, 8, 4, 7, 6, 3]
// console.log(pivot(arr));


const pivot = (arr, start=0, end=arr.length -1) =>{
    let pivot = arr[start]
    let swapIdx = start;

    for(let i = start + 1; i <= end; i++){
        if(pivot > arr[i]) {
            swapIdx++;
            [arr[i], arr[swapIdx]] = [arr[swapIdx], arr[i]];
        }
    }
    
    [arr[start], arr[swapIdx]] = [arr[swapIdx], arr[start]]
    // console.log(arr, '---- ');
    return swapIdx;
}

// console.log(pivot([4,8,2,1,5,7,6,3]));

const quickSort = (arr, left=0, right=arr.length -1)=>{

   if(left < right){
    const pivotIndex = pivot(arr,left,right);

    quickSort(arr,left,pivotIndex - 1);
    quickSort(arr,pivotIndex+1,right)

   }

   return arr

}

const arr = [5, 2, 1, 8, 4, 7, 6, 3]

// [2,1,4,8,23,67,88,53,45]
// [...]
// 1,2,8,4,23
//1,2,4,8,23

console.log(quickSort(arr));

File: sorting/IntermediateSort/quick_sort.py
def quick_sort(array, left=0, right=None):
    if right is None:
        right = len(array) - 1

    if left < right:
        pivot_point = pivot(array, left, right)
        quick_sort(array, left, pivot_point - 1)
        quick_sort(array, pivot_point + 1, right)

    return array


def pivot(item, start=0, end=None):
    if end is None:
        end = len(item)
    print(end)
    pivot_value = item[start]
    swap_index = start
    for i in range(start + 1, end + 1):
        if pivot_value > item[i]:
            swap_index += 1
            item[i], item[swap_index] = item[swap_index], item[i]
    item[start], item[swap_index] = item[swap_index], item[start]
    return swap_index


arr = [5, 2, 1, 8, 4, 7, 6, 3]

# 3, 2 , 1, 4, 5

# quick_sort(arr)
# print(pivot(arr))
quick_sort(arr)
print(arr)
